[RFE] Check whether Inspektor Gadget works on Mariner

## Current situation

Mariner is not an officially supported target platform and I don't know if it works.

## Impact

Users running Mariner are unsure whether they can run IG.

## Ideal future situation

- IG works on Mariner
- IG is documented to work on Mariner
- future releases of IG and Mariner maintain this compatibility.

## Implementation options

Test and talk with Mariner team if symbols needed.

## Additional information

I will also intro to Mariner PM.

Hi.


I deployed k8s using `minikube start --driver=docker` on [ISO](https://aka.ms/mariner-1.0-x86_64-iso) provided by CBL Mariner developers.

I needed to install `kernel-devel` package and then to copy `/usr/src/` into `gadget` container to avoid `bcc` gadgets complaining about missing kernel headers.

Then, I ran our integration tests, all tests are OK except the following:

```bash
--- FAIL: TestNetworkpolicy (51.39s)
    command.go:126: command: kubectl run --restart=Never --image=busybox -n test-ns test-pod -- sh -c 'while true; do wget -q -O /dev/null https://kinvolk.io; done'
    command.go:130: command returned:
        pod/test-pod created
        
    command.go:126: command: kubectl wait pod --for condition=ready -n test-ns test-pod
    command.go:130: command returned:
        pod/test-pod condition met
        
    command.go:126: command: $KUBECTL_GADGET network-policy monitor -n test-ns --output ./networktrace.log & sleep 15; kill $!; head networktrace.log
    command.go:130: command returned:
        Error detected: Error running command: command terminated with exit code 1
        
        /bin/sh: line 0: kill: (257951) - No such process
        failed to init perf map for IPv4 events: error guessing offsets: invalid guessing state while guessing source address, got checking expected checked
        
    command.go:138: regexp didn't match: "type":"connect".*"test-ns".*"test-pod"
        Error detected: Error running command: command terminated with exit code 1
        
        /bin/sh: line 0: kill: (257951) - No such process
        failed to init perf map for IPv4 events: error guessing offsets: invalid guessing state while guessing source address, got checking expected checked
--- FAIL: TestProcessCollector (8.02s)
    command.go:126: command: kubectl run --restart=Never --image=nginx -n test-ns test-pod
    command.go:130: command returned:
        pod/test-pod created
        
    command.go:126: command: kubectl wait pod --for condition=ready -n test-ns test-pod
    command.go:130: command returned:
        pod/test-pod condition met
        
    command.go:126: command: $KUBECTL_GADGET process-collector -n test-ns
    command.go:130: command returned:
        Failed to run the gadget on node "minikube": failed to create BPF collection: program dump_task: CO-RE relocations: no BTF for kernel version 5.10.78.1-2.cm1: not supported
        Failed to run the gadget on all nodes: None of them succeeded
        
    command.go:132: exit status 1
    command.go:126: command: kubectl delete pod -n test-ns test-pod
    command.go:130: command returned:
        pod "test-pod" deleted
        
FAIL
--- FAIL: TestSeccompadvisor (74.55s)
    command.go:126: command: kubectl run --restart=Never --image=busybox -n test-ns test-pod -- sh -c 'while true; do echo foo; done'
    command.go:130: command returned:
        pod/test-pod created
        
    command.go:126: command: kubectl wait pod --for condition=ready -n test-ns test-pod
    command.go:130: command returned:
        pod/test-pod condition met
        
    command.go:126: command: id=$($KUBECTL_GADGET seccomp-advisor start -n test-ns -p test-pod); sleep 30; $KUBECTL_GADGET seccomp-advisor stop $id
    command.go:130: command returned:
        Failed to run the gadget on node "minikube": Failed to start seccomp tracer: failed to create BPF collection: program tracepoint__raw_syscalls__sys_enter: CO-RE relocations: no BTF for kernel version 5.10.78.1-2.cm1: not supported
        Error: Failed to run the gadget on all nodes: None of them succeeded
        Error: usage: kubectl gadget seccomp-advisor stop global-trace-id
        
    command.go:132: exit status 1
    command.go:126: command: kubectl delete pod -n test-ns test-pod
    command.go:130: command returned:
        pod "test-pod" deleted
        
FAIL
```

For `network-policy`, I already had this error while trying `inspektor-gadget` on arm.
Sadly, I do not know the root cause of it but I will investigate.
For `*-collector` and `seccomp-advisor`, I think I can try to deploy `gadget` pod with `btfgen.sh` or adding some kernel options should to the trick.


Best regards.

Hi.

I built a [2.0-stable]() Mariner ISO which I installed through `qemu`.
All the gadgets are working expect the following:

```
--- FAIL: TestTraceloop (43.63s)
    command.go:101: command: kubectl create ns test-traceloop-2933568871211445515
    command.go:105: command returned:
        namespace/test-traceloop-2933568871211445515 created
        
    command.go:101: command: $KUBECTL_GADGET traceloop start
    command.go:105: command returned:
        
    command.go:101: command: sleep 15
    command.go:105: command returned:
        
    command.go:101: command: kubectl run --restart=Never -n test-traceloop-2933568871211445515 --image=busybox multiplication -- sh -c 'RANDOM=ou
tput ; echo "3*7*2" | bc > /tmp/file-$RANDOM ; sleep infinity'
    command.go:105: command returned:
        pod/multiplication created
        
    command.go:101: command: sleep 5 ; kubectl wait -n test-traceloop-2933568871211445515 --for=condition=ready pod/multiplication ; kubectl get 
pod -n test-traceloop-2933568871211445515 ; sleep 2
    command.go:105: command returned:
        pod/multiplication condition met
        NAME             READY   STATUS    RESTARTS   AGE
        multiplication   1/1     Running   0          5s
        
    command.go:101: command: sleep 20 ; $KUBECTL_GADGET traceloop list -n test-traceloop-2933568871211445515 --no-headers | grep multiplication |
 awk '{print $1" "$6}'
    command.go:105: command returned:
        time="2022-02-02T16:18:38Z" level=fatal msg="Error in getting traces: \"Traceloop is not running on any node. Please start it with 'kubec
tl gadget traceloop start'.\"" args="[]" command="kubectl-gadget traceloop list"
        
    command.go:117: diff: ["multiplication started\n" != "time=\"2022-02-02T16:18:38Z\" level=fatal msg=\"Error in getting traces: \\\"Traceloop 
is not running on any node. Please start it with 'kubectl gadget traceloop start'.\\\"\" args=\"[]\" command=\"kubectl-gadget traceloop list\"\n"
]
    command.go:101: command: $KUBECTL_GADGET traceloop list -A
    command.go:105: command returned:
        time="2022-02-02T16:18:38Z" level=fatal msg="Error in getting traces: \"Traceloop is not running on any node. Please start it with 'kubec
tl gadget traceloop start'.\"" args="[]" command="kubectl-gadget traceloop list"
        
    command.go:107: exit status 1

--- FAIL: TestTcptop (60.53s)
    command.go:101: command: kubectl create ns test-tcptop-3510942875414458836
    command.go:105: command returned:
        namespace/test-tcptop-3510942875414458836 created
        
    command.go:151: Start command: $KUBECTL_GADGET tcptop --node $(kubectl get node --no-headers | cut -d' ' -f1) -n test-tcptop-3510942875414458
836 -p test-pod
    command.go:101: command: kubectl run --restart=Never --image=busybox -n test-tcptop-3510942875414458836 test-pod -- sh -c 'while true; do wge
t -q -O /dev/null https://kinvolk.io; done'
    command.go:105: command returned:
        pod/test-pod created
        
    command.go:101: command: kubectl wait pod --for condition=ready -n test-tcptop-3510942875414458836 test-pod
    16:19:11 loadavg: 2.53 1.60 1.98 3/709 324008
        
        
    command.go:101: command: kubectl delete ns test-tcptop-3510942875414458836
    command.go:105: command returned:
        namespace "test-tcptop-3510942875414458836" deleted
        
--- FAIL: TestSeccompadvisor (80.10s)
    command.go:101: command: kubectl create ns test-seccomp-advisor-4751997750760398084
    command.go:105: command returned:
        namespace/test-seccomp-advisor-4751997750760398084 created
        
    command.go:101: command: kubectl run --restart=Never --image=busybox -n test-seccomp-advisor-4751997750760398084 test-pod -- sh -c 'while tru
e; do echo foo; done'
    command.go:105: command returned:
        pod/test-pod created
        
    command.go:101: command: kubectl wait pod --for condition=ready -n test-seccomp-advisor-4751997750760398084 test-pod
    command.go:105: command returned:
        pod/test-pod condition met
        
    command.go:101: command: id=$($KUBECTL_GADGET seccomp-advisor start -n test-seccomp-advisor-4751997750760398084 -p test-pod); sleep 30; $KUBE
CTL_GADGET seccomp-advisor stop $id
    command.go:105: command returned:
        
    command.go:113: regexp didn't match: write
        
    command.go:101: command: kubectl delete ns test-seccomp-advisor-4751997750760398084
    command.go:105: command returned:
        namespace "test-seccomp-advisor-4751997750760398084" deleted
        
--- FAIL: TestNetworkpolicy (66.43s)
    command.go:101: command: kubectl create ns test-networkpolicy-605394647632969758
    command.go:105: command returned:
        namespace/test-networkpolicy-605394647632969758 created
        
    command.go:101: command: kubectl run --restart=Never --image=busybox -n test-networkpolicy-605394647632969758 test-pod -- sh -c 'while true; 
do wget -q -O /dev/null https://kinvolk.io; done'
    command.go:105: command returned:
        pod/test-pod created
        
    command.go:101: command: kubectl wait pod --for condition=ready -n test-networkpolicy-605394647632969758 test-pod
    command.go:105: command returned:
        pod/test-pod condition met
        
    command.go:101: command: $KUBECTL_GADGET network-policy monitor -n test-networkpolicy-605394647632969758 --output ./networktrace.log & sleep 
15; kill $!; head networktrace.log
    command.go:105: command returned:
        Error detected: Error running command: command terminated with exit code 1
        
        /bin/sh: line 1: kill: (350276) - No such process
        failed to init perf map for IPv4 events: error guessing offsets: invalid guessing state while guessing source address, got checking expec
ted checked
        
    command.go:113: regexp didn't match: "type":"connect".*"test-networkpolicy-605394647632969758".*"test-pod"
        Error detected: Error running command: command terminated with exit code 1
        
        /bin/sh: line 1: kill: (350276) - No such process
        failed to init perf map for IPv4 events: error guessing offsets: invalid guessing state while guessing source address, got checking expec
ted checked
        
    command.go:101: command: kubectl delete ns test-networkpolicy-605394647632969758
    command.go:105: command returned:
        namespace "test-networkpolicy-605394647632969758" deleted
```

I tested them again and sadly I am not able, for the moment, to tell exactly while they are failling:
1. Absolutely nothing is printed for `tcptop`.
2. The same for `seccomp-advisor`.
3. `traceloop` complained about being already run which is not true.
4. `network-policy` already showed this error.

Note that `tcptop` was working on 1.0 but not on 2.0.

Bes regards.

--- a/cmd/kubectl-gadget/deploy.go
+++ b/cmd/kubectl-gadget/deploy.go
@@ -291,7 +291,8 @@ spec:
               # https://elixir.bootlin.com/linux/v5.14.14/source/kernel/ptrace.c#L284
               - SYS_PTRACE
 
-              # Needed by setrlimit in gadgettracermanager.
+              # Needed by setrlimit in gadgettracermanager and by the traceloop
+              # gadget.
               - SYS_RESOURCE
 
               # Needed for gadgets that don't dumb the memory rlimit.

--- a/pkg/gadgets/traceloop/gadget.go
+++ b/pkg/gadgets/traceloop/gadget.go
@@ -121,6 +121,13 @@ func (t *Trace) Start(trace *gadgetv1alpha1.Trace) {
 		echo "-:pfree_uts_ns" >> /sys/kernel/debug/tracing/kprobe_events 2>/dev/null || true
 		echo "-:pcap_capable" >> /sys/kernel/debug/tracing/kprobe_events 2>/dev/null || true
 
+		# Remove the opened files limit to avoid getting this error:
+		# error while loading "tracepoint/raw_syscalls/sys_enter" (too many open files):
+		# Indeed, traceloop creates a lof of tracers which each has maps and events:
+		# https://github.com/kinvolk/traceloop/blob/6f4efc6fca46d92c75f4ec4e6c6e1d829bdeaddf/bpf/straceback-guess-bpf.h#L27-L28
+		# So, this can generate a lof ot opened files.
+		ulimit -n unlimited
+
 		rm -f /run/traceloop.socket
 		exec /bin/traceloop k8s
 	`)

